clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1721,42 +1722,48 @@ void handleFailure(Throwable throwable) {
     public CreateAclsResult createAcls(Collection<AclBinding> acls, CreateAclsOptions options) {
         final long now = time.milliseconds();
         final Map<AclBinding, KafkaFutureImpl<Void>> futures = new HashMap<>();
-        final List<AclCreation> aclCreations = new ArrayList<>();
+        final List<CreateAclsRequestData.CreatableAcl> aclCreations = new ArrayList<>();
         for (AclBinding acl : acls) {
             if (futures.get(acl) == null) {
                 KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
                 futures.put(acl, future);
                 String indefinite = acl.toFilter().findIndefiniteField();
                 if (indefinite == null) {
-                    aclCreations.add(new AclCreation(acl));
+                    aclCreations.add(CreateAclsRequest.creatableAcl(acl));
                 } else {
                     future.completeExceptionally(new InvalidRequestException("Invalid ACL creation: " +
                         indefinite));
                 }
             }
         }
+        final CreateAclsRequestData data = new CreateAclsRequestData().setCreations(aclCreations);
         runnable.call(new Call("createAcls", calcDeadlineMs(now, options.timeoutMs()),
             new LeastLoadedNodeProvider()) {
 
             @Override
             CreateAclsRequest.Builder createRequest(int timeoutMs) {
-                return new CreateAclsRequest.Builder(aclCreations);
+                return new CreateAclsRequest.Builder(data);
             }
 
             @Override
             void handleResponse(AbstractResponse abstractResponse) {
                 CreateAclsResponse response = (CreateAclsResponse) abstractResponse;
-                List<AclCreationResponse> responses = response.aclCreationResponses();
-                Iterator<AclCreationResponse> iter = responses.iterator();
-                for (AclCreation aclCreation : aclCreations) {
-                    KafkaFutureImpl<Void> future = futures.get(aclCreation.acl());
+                List<CreateAclsResponseData.CreatableAclResult> responses = response.results();
+                Iterator<CreateAclsResponseData.CreatableAclResult> iter = responses.iterator();
```

Comment:
This part isn't quite right.  We should be iterating over `aclCreations`, since that is what we actually sent out on the wire.  Entries in `acls` may or may not have been sent out on the wire (if they were invalid, they weren't.)

The TODO comment about "duplicates" is misguided and should be removed.  Since duplicates never get sent on the wire, they don't need to be handled here.

Reply:
I think you assumed this was complete, it's WIP as the commit message says: https://github.com/apache/kafka/pull/7725/commits/755a6f2c7273ac570bacbdbe0c2a1a8e2dc94088

The TODO is also a hint of the same. :)